File: dnaviz.py
from squiggle import transform
import time
import statistics
import string
import random
def randomSeq(size, chars):
    return ''.join(random.choice(chars) for _ in range(size))

E_4 = randomSeq(10000, 'ATGC')
E_5 = randomSeq(100000, 'ATGC')
E_6 = randomSeq(1000000, 'ATGC')

methods = ['squiggle', 'yau', 'yau-bp', 'gates', 'randic', 'qi']

for method in methods:
    i = 0
    times_squ_4 = []
    while i < 1000:
        t0 = time.time()
        transform(E_4, method = method)
        t1 = time.time()
        times_squ_4.append(t1 - t0)
        i += 1

    # print(method + ' 10,000 median: ' + str(statistics.median(times_squ_4)) + 's')
    # print(method + ' 10,000 mean: ' + str(statistics.mean(times_squ_4)) + 's')
    # print(method + ' 10,000 standard deviation: ' + str(statistics.stdev(times_squ_4)) + 's')
    print('| ' + method + ' | ' + '10,000 | ' +  str(statistics.median(times_squ_4)) + ' | ' + str(statistics.mean(times_squ_4)) + ' | ' + str(statistics.stdev(times_squ_4)) + ' |')

for method in methods:
    i = 0
    times_squ_5 = []
    while i < 1000:
        t0 = time.time()
        transform(E_5, method = method)
        t1 = time.time()
        times_squ_5.append(t1 - t0)
        i += 1

    # print(method + ' 100,000 median: ' + str(statistics.median(times_squ_5)) + 's')
    # print(method + ' 100,000 mean: ' + str(statistics.mean(times_squ_5)) + 's')
    # print(method + ' 100,000 standard deviation: ' + str(statistics.stdev(times_squ_5)) + 's')
    print('| ' + method + ' | ' + '100,000 | ' +  str(statistics.median(times_squ_5)) + ' | ' + str(statistics.mean(times_squ_5)) + ' | ' + str(statistics.stdev(times_squ_5)) + ' |')

for method in methods:
    i = 0
    times_squ_6 = []
    while i < 1000:
        t0 = time.time()
        transform(E_6, method = method)
        t1 = time.time()
        times_squ_6.append(t1 - t0)
        i += 1

    # print(method + ' 1,000,000 median: ' + str(statistics.median(times_squ_6)) + 's')
    # print(method + ' 1,000,000 mean: ' + str(statistics.mean(times_squ_6)) + 's')
    # print(method + ' 1,000,000 standard deviation: ' + str(statistics.stdev(times_squ_6)) + 's')
    print('| ' + method + ' | ' + '1,000,000 | ' +  str(statistics.median(times_squ_6)) + ' | ' + str(statistics.mean(times_squ_6)) + ' | ' + str(statistics.stdev(times_squ_6)) + ' |')

File: dnaviz.ts
import * as dna from "dnaviz";
import * as ss from "simple-statistics";
function randomSeq(length) {
  var result = '';
  var characters = 'ATGC';
  var charactersLength = characters.length;
  for ( var i = 0; i < length; i++ ) {
     result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}
const E_4 = randomSeq(10000)
const E_5 = randomSeq(100000)
const E_6 = randomSeq(1000000)
// array of all visualization methods so each test can iterate over each
const methods = [
  dna.squiggle,
  dna.yau,
  dna.yau_bp,
  dna.gates,
  dna.randic,
  dna.qi,
];

// tests 1000 times for each method
for (let i = 0; i < methods.length; i++) {
  let j = 0;
  const times_squ_4 = [];
  while (j < 1000) {
    let then = new Date();
    methods[i](E_4);
    let now = new Date();
    times_squ_4.push((now.getTime() - then.getTime()) / 1000);
    j++;
  }

  console.log(methods[i].name + " 10,000 median: " + ss.median(times_squ_4) + "s");
  console.log(methods[i].name + " 10,000 mean: " + ss.mean(times_squ_4) + "s");
  console.log(
    methods[i].name + " 10,000 standard dev: " + ss.standardDeviation(times_squ_4) + "s"
  );
}

for (let i = 0; i < methods.length; i++) {
  let j = 0;
  const times_squ_5 = [];
  while (j < 1000) {
    let then = new Date();
    methods[i](E_5);
    let now = new Date();
    times_squ_5.push((now.getTime() - then.getTime()) / 1000);
    j++;
  }

  console.log(methods[i].name + " 100,000 median: " + ss.median(times_squ_5) + "s");
  console.log(methods[i].name + " 100,000 mean: " + ss.mean(times_squ_5) + "s");
  console.log(
    methods[i].name + " 100,000 standard dev: " + ss.standardDeviation(times_squ_5) + "s"
  );
}

for (let i = 0; i < methods.length; i++) {
  let j = 0;
  const times_squ_6 = [];
  while (j < 1000) {
    let then = new Date();
    methods[i](E_6);
    let now = new Date();
    times_squ_6.push((now.getTime() - then.getTime()) / 1000);
    j++;
  }

  console.log(methods[i].name + " 1,000,000 median: " + ss.median(times_squ_6) + "s");
  console.log(methods[i].name + " 1,000,000 mean: " + ss.mean(times_squ_6) + "s");
  console.log(
    methods[i].name + " 1,000,000 standard dev: " + ss.standardDeviation(times_squ_6) + "s"
  );
}
